extra_data/filecache.py

```diff
@@ -0,0 +1,125 @@
+"""
+File cache implementation
+
+File cache provides the collection for open h5 files 
+which keeps the number of opened files under the given limit
+popping the files that were accessed the longest time ago
+
+Implementation based on the OrderedDict
+
+The module provides global instance of file cache for using
+across entire process
+"""
+import h5py
+from collections import OrderedDict
+
+class FileRef(object):
+    __slots__ = 'nref', 'fh'
+
+
+class FileCache(object):
+    """
+    FileCache is a collection for opened h5 files. It keeps the number of opened files 
+    under the given limit popping files accessed longest time ago.
+    """
+    def __init__(self, maxfiles=128):
+        self._maxfiles = maxfiles
+        self._cache = OrderedDict()
+        
+    def __del__(self):
+        self.clean()
+        
+    @property
+    def maxfiles(self):
+        return self._maxfiles
+    
+    @maxfiles.setter
+    def maxfiles(self, maxfiles):
+        """
+        Set the new file limit and pops redundant files
+        """
+        n = len(self._cache)
+        while n > maxfiles:
+            p, f = self._cache.popitem(last=False)
+            f.fh.close()
+            n -= 1
+        self._maxfiles = maxfiles
+
+    def clean(self):
+        """
+        Closes all cached files
+        """
+        while len(self._cache):
+            p, f = self._cache.popitem(last=False)
+            f.fh.close()
+
+    def open(self, filename):
+        """
+        Returns the opened `h5.File` instance from the cache.
+        It opens new file only if the requested file is absent.
+        If new file exceeds the *maxfiles* limit it pops file accessed most far.
+
+        For use of the file cache, FileAccess shold use `get_or_open(filename)`
+        instead of direct opening file with h5py
+        """
+        try:
+            f = self._cache[filename]
+            f.nref += 1
+            self._cache.move_to_end(filename)
+        except KeyError:
+            if len(self._cache) >= self._maxfiles:
+                rp, rf = self._cache.popitem(last=False)
+                rf.fh.close()
+            f = FileRef()
+            f.fh = h5py.File(filename, 'r')
+            f.nref = 1
```

Comment:
I'm inclined to give `FileRef` a `__init__()` method so these three lines are one. It may seem pointless, but it's nice to have objects which can't be left half initialised.

Reply:
I will fix.